File: money.rb
require_relative 'expression'

class Money < Expression

  attr_reader :currency

  def initialize (amount, currency)
    @amount   = amount
    @currency = currency
  end

  def == (money)
    @amount == money.amount && @currency == money.currency
  end

  def times (multiplier)
    Money.new (@amount * multiplier), @currency
  end

  def plus (addend)
    Sum.new self, addend
  end

  def self.dollar(amount)
    Money.new amount, 'USD'
  end

  def self.franc(amount)
    Money.new amount, 'CHF'
  end

  def inspect
    "#{@amount} #{@currency}"
  end

  def reduce (bank, to)
    rate = bank.rate currency, to
    Money.new amount / rate, to
  end

  attr_reader :amount

  protected

  @amount   = 0
  @currency = ''

end

File: Money.java

class Money implements Expression {

  static Money dollar(int amount) {
    return new Money(amount, "USD");
  }

  static Money franc(int amount) {
    return new Money(amount, "CHF");
  }

  protected int amount;
  protected String currency;

  Money(int amount, String currency) {
    this.amount = amount;
    this.currency = currency;
  }

  public boolean equals(Object object) {
    Money money = (Money) object;
    return amount == money.amount
      && currency().equals(money.currency());
  }

  Money times(int multiplier) {
    return new Money(amount * multiplier, currency);
  }

  Expression plus(Expression addend) {
    return new Sum(this, addend);
  }

  String currency() {
    return currency;
  }

  public String toString() {
    return amount + " " + currency;
  }

  public Money reduce(Bank bank, String to) {
    int rate = bank.rate(currency, to);
    return new Money(amount / rate, to);
  }
}
